diff --git a/CMakeLists.txt b/CMakeLists.txt
index d8336b132..8bf4eea97 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -667,7 +667,6 @@ if(ENABLE_TBB_DEVICE)
   # related documentation: https://github.com/oneapi-src/oneTBB/tree/tbb_2020/cmake#binary-package-integration
   find_package(TBB REQUIRED)
   if(TARGET TBB::tbb)
-    message(STATUS "Found TBB::tbb target")
     set(TBB_IMPORTED_TARGETS TBB::tbb)
   elseif(NOT TBB_FOUND AND PKG_CONFIG_FOUND)
     message(STATUS "searching for TBB with pkg-config")
@@ -683,11 +682,8 @@ if(ENABLE_TBB_DEVICE)
   if(NOT TBB_LIBRARIES)
     get_target_property(TBB_LIBRARIES TBB::tbb IMPORTED_LOCATION_RELEASE)
   endif()
-  if(NOT TBB_LIBRARIES)
-    get_target_property(TBB_LIBRARIES TBB::tbb IMPORTED_LOCATION_NONE)
-  endif()
 
-  if((NOT TBB_LIBRARIES) OR (NOT TBB_INCLUDE_DIRS))
+  if(NOT TBB_FOUND)
     message(FATAL_ERROR "Can't find TBB libraries")
   endif()