commit 4575fe09bb661470eca4dc934a982b77c0f06294
Author: Remi Herilier <remi.herilier@picviz.com>
Date:   Mon Feb 3 11:40:18 2014 +0100

    00-debian-changelog.patch

diff --git a/debian/changelog b/debian/changelog
index e99046a..79a1b0e 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,9 @@
+qt4-x11 (4:4.8.5+git209-g718fae5+dfsg-1.1) unstable; urgency=medium
+
+  * apply Picviz's patches
+
+ -- Remi Herilier <rherilier@picviz.com>  Mon, 03 Feb 2014 11:39:58 +0100
+
 qt4-x11 (4:4.8.5+git209-g718fae5+dfsg-1) unstable; urgency=medium
 
   [ Dmitry Shachnev ]
