diff --git a/src/gui/graphicsview/qgraphicsscene.cpp b/src/gui/graphicsview/qgraphicsscene.cpp
index 5c31227..a566d70 100644
--- a/src/gui/graphicsview/qgraphicsscene.cpp
+++ b/src/gui/graphicsview/qgraphicsscene.cpp
@@ -1091,7 +1091,7 @@ QList<QGraphicsItem *> QGraphicsScenePrivate::itemsAtPosition(const QPoint &scre
     Q_Q(const QGraphicsScene);
     QGraphicsView *view = widget ? qobject_cast<QGraphicsView *>(widget->parentWidget()) : 0;
     if (!view)
-        return q->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder, QTransform());
+        return q->items(scenePos, Qt::IntersectsItemShape, Qt::DescendingOrder, defaultViewTransform);
 
     const QRectF pointRect(QPointF(widget->mapFromGlobal(screenPos)), QSizeF(1, 1));
     if (!view->isTransformed())
@@ -5637,6 +5637,29 @@ void QGraphicsScene::setPalette(const QPalette &palette)
     d->setPalette_helper(resolvedPalette);
 }
 
+/*! \since 4.8.4 (Picviz)
+ *
+ * Set the default scene-to-view transformation used if none can be found.
+ * By default, an identity matrix is used.
+ */
+void QGraphicsScene::setDefaultViewTransform(QTransform const& viewTransform)
+{
+	Q_D(QGraphicsScene);
+	d->defaultViewTransform = viewTransform;
+}
+
+/*! \since 4.8.4 (Picviz)
+ *
+ * Get the default scene-to-view transformation used if none can be found.
+ *
+ * \sa setDefaultTransform
+ */
+QTransform const& QGraphicsScene::defaultViewTransform() const
+{
+	Q_D(const QGraphicsScene);
+	return d->defaultViewTransform;
+}
+
 /*!
     \since 4.6
 
diff --git a/src/gui/graphicsview/qgraphicsscene.h b/src/gui/graphicsview/qgraphicsscene.h
index cf221c0..2d473d7 100644
--- a/src/gui/graphicsview/qgraphicsscene.h
+++ b/src/gui/graphicsview/qgraphicsscene.h
@@ -251,6 +251,9 @@ public:
 
     bool sendEvent(QGraphicsItem *item, QEvent *event);
 
+	void setDefaultViewTransform(QTransform const& viewTransform);
+	QTransform const& defaultViewTransform() const;
+
 public Q_SLOTS:
     void update(const QRectF &rect = QRectF());
     void invalidate(const QRectF &rect = QRectF(), SceneLayers layers = AllLayers);
diff --git a/src/gui/graphicsview/qgraphicsscene_p.h b/src/gui/graphicsview/qgraphicsscene_p.h
index 0911f3d..958d2b1 100644
--- a/src/gui/graphicsview/qgraphicsscene_p.h
+++ b/src/gui/graphicsview/qgraphicsscene_p.h
@@ -71,6 +71,7 @@
 #include <QtGui/qpalette.h>
 #include <QtGui/qstyle.h>
 #include <QtGui/qstyleoption.h>
+#include <QtGui/qtransform.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -180,6 +181,8 @@ public:
     QList<QGraphicsItem *> cachedItemsUnderMouse;
     QList<QGraphicsItem *> hoverItems;
     QPointF lastSceneMousePos;
+
+	QTransform defaultViewTransform;
     void enableMouseTrackingOnViews();
     QMap<Qt::MouseButton, QPointF> mouseGrabberButtonDownPos;
     QMap<Qt::MouseButton, QPointF> mouseGrabberButtonDownScenePos;
