diff --git a/debian/libqt4-dev.install b/debian/libqt4-dev.install
index 0a0245c..fb37adb 100644
--- a/debian/libqt4-dev.install
+++ b/debian/libqt4-dev.install
@@ -2860,6 +2860,13 @@ usr/include/qt4/QtXmlPatterns/qxmlresultitems.h
 usr/include/qt4/QtXmlPatterns/qxmlschema.h
 usr/include/qt4/QtXmlPatterns/qxmlschemavalidator.h
 usr/include/qt4/QtXmlPatterns/qxmlserializer.h
+usr/include/qt4/QtGui/qscrollbar64.h
+usr/include/qt4/QtGui/QAbstractSlider64
+usr/include/qt4/QtGui/qabstractslider64.h
+usr/include/qt4/QtGui/QStyleOptionSlider64
+usr/include/qt4/QtGui/QScrollBar64
+usr/include/qt4/Qt/qscrollbar64.h
+usr/include/qt4/Qt/qabstractslider64.h
 usr/lib/*/libQt3Support.prl
 usr/lib/*/libQt3Support.so
 usr/lib/*/libQtCLucene.prl
