diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 41b2a28..f11dce2 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -13,6 +13,48 @@ target_include_directories(remoteexecution
     PRIVATE
     nlohmann_json::nlohmann_json
 )
+
+
+find_package(utf8_range REQUIRED)
+find_package(absl REQUIRED)
+
+ set(protobuf_ABSL_USED_TARGETS
+    absl::absl_check
+    absl::absl_log
+    absl::algorithm
+    absl::base
+    absl::bind_front
+    absl::bits
+    absl::btree
+    absl::cleanup
+    absl::cord
+    absl::core_headers
+    absl::debugging
+    absl::die_if_null
+    absl::dynamic_annotations
+    absl::flags
+    absl::flat_hash_map
+    absl::flat_hash_set
+    absl::function_ref
+    absl::hash
+    absl::layout
+    absl::log_initialize
+    absl::log_severity
+    absl::memory
+    absl::node_hash_map
+    absl::node_hash_set
+    absl::optional
+    absl::span
+    absl::status
+    absl::statusor
+    absl::strings
+    absl::synchronization
+    absl::time
+    absl::type_traits
+    absl::utility
+    absl::variant
+  )
+
 target_link_libraries(remoteexecution
     ${_EXTRA_LDD_FLAGS}
     Buildbox::buildboxcommon
@@ -22,6 +64,8 @@ target_link_libraries(remoteexecution
     ${STATIC_GRPC_LINKER_RULE}
     ${ZLIB_LIBRARIES}
     ${OS_LIBS}
+    ${protobuf_ABSL_USED_TARGETS}
+    utf8_range::utf8_validity
 )
 
 if (CMAKE_SYSTEM_NAME MATCHES "SunOS")
